src/nvim/liveupdate.c

```diff
@@ -0,0 +1,184 @@
+#include "nvim/liveupdate.h"
+#include "nvim/memline.h"
+#include "nvim/api/private/helpers.h"
+#include "nvim/msgpack_rpc/channel.h"
+
+/*
+ * Register a channel. Return True if the channel was added, or already added.
+ * Return False if the channel couldn't be added because the buffer is
+ * unloaded.
+ */
+bool liveupdate_register(buf_T *buf, uint64_t channel_id) {
+  // must fail if the buffer isn't loaded
+  if (buf->b_ml.ml_mfp == NULL) {
+    return false;
+  }
+
+  // count how many channels are currently watching the buffer
+  int active_size = 0;
+  if (buf->liveupdate_channels != NULL) {
+    while (buf->liveupdate_channels[active_size] != LIVEUPDATE_NONE) {
+      if (buf->liveupdate_channels[active_size] == channel_id) {
+        // buffer is already registered ... nothing to do
+        return true;
+      }
+      active_size++;
+    }
+  }
+
+  // add the buffer to the active array and send the full buffer contents now
+  uint64_t *newlist = xmalloc((active_size + 2) * sizeof channel_id);
```

Comment:
I've made a new commit that changes `buf->liveupdate_channels` to a `kvec_t(uint64_t)`.